File: modules/core/src/main/java/turbolift/internals/engine/concurrent/atomic/AtomicIntVH.java
package turbolift.internals.engine.concurrent.atomic;
import java.lang.invoke.VarHandle;
import java.lang.invoke.MethodHandles;


public class AtomicIntVH {
  private int valueVH;
  private static final VarHandle nameVH;

  static {
    try {
      nameVH = MethodHandles.lookup().findVarHandle(AtomicIntVH.class, "valueVH", int.class);
    } catch (ReflectiveOperationException e) {
      throw new ExceptionInInitializerError(e);
    }
  }

  public AtomicIntVH(int a) {
    valueVH = a;
  }

  protected int getVH() {
    return (int) nameVH.get(this);
  }

  protected void setVH(int a) {
    nameVH.set(this, a);
  }

  protected int gasVH(int a) {
    return (int) nameVH.getAndSet(this, a);
  }

  protected boolean casVH(int a, int b) {
    return nameVH.compareAndSet(this, a, b);
  }

  protected int caxVH(int a, int b) {
    return (int) nameVH.compareAndExchange(this, a, b);
  }

  protected int gaaVH(int d) {
    return (int) nameVH.getAndAdd(this, d);
  }
}

File: modules/core/src/main/scala/turbolift/internals/engine/concurrent/atomic/AtomicIntVH.scala
package turbolift.internals.engine.concurrent.atomic
import java.lang.invoke.{VarHandle, MethodHandles}


private[turbolift] class AtomicIntVH(protected var unsafeValue: Int):
  protected def getVH: Int = AtomicIntVH.theVH.get(this)
  
  protected def setVH(a: Int): Unit = AtomicIntVH.theVH.set(this, a)
  
  protected def gasVH(a: Int): Int = AtomicIntVH.theVH.getAndSet(this, a)

  protected def casVH(a: Int, b: Int): Boolean = AtomicIntVH.theVH.compareAndSet(this, a, b)

  protected def caxVH(a: Int, b: Int): Int = AtomicIntVH.theVH.compareAndExchange(this, a, b)


private[turbolift] object AtomicIntVH:
  private val theVH: VarHandle = MethodHandles
    .privateLookupIn(classOf[AtomicIntVH], MethodHandles.lookup()).nn
    .findVarHandle(classOf[AtomicIntVH], "unsafeValue", java.lang.Integer.TYPE).nn
